File: apps/server2/src/api/components/health_check/health_check_service.py
from abc import ABC, abstractmethod

from fastapi import status

from server_error import Detail, ServerError
from services.db_service import DBService


class IHealthCheckService(ABC):
    @abstractmethod
    async def check_health(self) -> bool:
        raise Exception("NotImplementedException")


class HealthCheckService(IHealthCheckService):
    def __init__(self, db_service: DBService):
        self.db_service = db_service

    async def check_health(self) -> bool:
        try:
            return await self.db_service.check_database_is_alive()
        except Exception as error:
            message = "An error occurred when checking if application is alive"
            print(message)
            raise ServerError(
                message,
                status.HTTP_500_INTERNAL_SERVER_ERROR,
                Detail(context="unknown", cause=str(error)),
            )

File: apps/server1/src/api/components/health-check/health-check-service.ts
import httpStatus from 'http-status'

import { ServerError } from '../../../server-error'
import { DBService } from '../../../services'

interface IHealthCheckService {
  checkHealth(): Promise<boolean | void>
}

class HealthCheckService implements IHealthCheckService {
  constructor(private dbService: DBService) {}

  async checkHealth(): Promise<boolean | void> {
    try {
      return await this.dbService.checkDatabaseIsAlive()
    } catch (err) {
      const message =
        'An error occurred when checking if application is healthy'
      throw new ServerError(message, httpStatus.INTERNAL_SERVER_ERROR, {
        context: undefined,
        cause: err instanceof ServerError ? err.newDetail?.cause : String(err),
      })
    }
  }
}

export { HealthCheckService, IHealthCheckService }
